### PR TITLE
add init mod notebook

### PR DESCRIPTION
**Pull request by MattGraz**
@chrfarri:

I was toying around with some modeling this afternoon and got some promising initial results. 

I trained models to predict each of the note (A, B, C...), the structure ([these](https://github.com/gciruelos/musthe/blob/0ac6b3cf46482ca438a6dfe94800c2082a5da3f3/musthe/musthe.py#L306)), and the instrument type (limited to guitar, strings, keyboard, and bass right now). 

For each I was able to get out of sample classification rates of 91%, 80%, and 98% respectively.  These came from a multi-output neural network output at the end of the notebook. 

Some thoughts on some todos:

# TO DOs:

- Generate chords for more instruments and see if we can still get high accuracy when we increase that.  The instruments we currently have sound fairly distinct so it is likely fairly "easy" for the model to achieve class separation.
- Potentially play with more instrument sources.  All these chords are from acoustic instrumentation, but we can generate data from the "synthetic" as well as "electronic" notes as well.
- Run model with all data and see if we can push accuracy further.  My training set has 80,000 observations, validation set 20,000 observations, and test set 10,000.
- Play around more with model architecture and see if that does anything.  I 100% arbitrarily selected the one I used. 
- Add more time to each note and vary the length (time) of the chords.  Currently, I only generated chords that are 1 second long (primarily because I am out of disk space :) ), but all the notes have up to 4 seconds to them.  In the models, I transform all the sounds from the time --> frequency domain.  It would be interesting to learn if we can train a model that is resilient to time length of the file and if that effects the frequency representation at all. 
- Generally speaking, the notes we have are fairly clean-ish.  It will be interesting if we add some of the recorded data you provide to the mix to see if the model can generalize beyond the nsynth dataset

### CODE REVIEW

**Comment by chrfarri** [comment]
Where is the modeling portion?

**Comment by MattGraz** [comment]
Might be easier to view the notebook at the following:
https://github.com/MattGraz/engr511/blob/init_model/notebooks/generated_chord_model.ipynb

There is a heading "Modeling" where I do the training

**Comment by chrfarri** [comment]
Ah, I think I was looking at an older version or something. I see it all now. Cool. Would you be able to meet tomorrow? I’d like to go over some of the modeling code and next steps.

It looks like you’re just using the STFT transformation as input. Pretty good results for no transformation of the data.

**Comment by MattGraz** [comment]
Yep - I think there is definitely some accuracy left on the table that we can capture.

**Comment by MattGraz** [comment]
@chrfarri : Family coming in this weekend but free anytime tonight, tomorrow around 4-5, or Sunday afternoon?

**Comment by chrfarri** [comment]
Tomorrow at 4 works for me.

**Comment by MattGraz** [comment]
One thing that came to mind is that these models have some data leakage.   Each instrument produces up to 5 of the same chord.  They are often VERY similar to one another (at least by my ear) with only subtle differences.

My guess is that one reason the results appear so good is that even though a true test/holdout set was used, more than likely the test set contains chords from instruments used in training.  As such, the test set data is largely correlated with what I used for train. 

The nsynth project provides train/validation/test datasets.  All the chords I generated came from the "Train" section.  I'll look into generating chords from the validation and test data which contains new instruments not in the training data.  This will give us a better sense of where we are at model wise.

**Comment by MattGraz** [comment]
I updated the notebook to have distinct instrument IDs across train/validation/test sets and we still get good performance :).  

I'll push the updates once the notebook is done running.  Will probably take the rest of the night.
